### PR TITLE
formatter: add text_format_source, relax minimum string length on text_format

### PR DESCRIPTION
**Pull request by esmet**
This change is motivated by https://github.com/envoyproxy/envoy/pull/14221 where we use a SubstitutionFormatString as a way to define custom HTTP response body rewrites.

Commit Message: formatter: add text_format_source, relax minimum string length on text_format in SubstitutionFormatString
Additional Description: The relaxed field validation on text_format now allows a user to replace something with nothing, e.g. to replace a non-empty HTTP response body with an empty one. The text_format_source field allows for a DataSource to be used to supply text inside of providing it inline.
Risk Level: low (new fields)
Testing: unit test needed for text_format_source
Docs Changes: NEEDED
Release Notes: NEEDED
Platform Specific Features:

### CODE REVIEW

File: api/envoy/config/core/v3/substitution_format_string.proto [hunk 1]
```diff
@@ -61,6 +64,12 @@ message SubstitutionFormatString {
     //  }
     //
     google.protobuf.Struct json_format = 2 [(validate.rules).message = {required: true}];
+
+    // Specify a format with command operators, similar to text_format, but from a DataSource.
+    DataSource text_format_source = 5;
+
+    // Specify fixed text with no command operators.
+    string fixed_text = 6;
```
Comment (qiwzhang) [inline api/envoy/config/core/v3/substitution_format_string.proto:72]: what is the purpose of this format?  Especially how it is used in file_access_log

File: api/envoy/config/core/v3/substitution_format_string.proto [hunk 2]
```diff
@@ -61,6 +64,12 @@ message SubstitutionFormatString {
     //  }
     //
     google.protobuf.Struct json_format = 2 [(validate.rules).message = {required: true}];
+
+    // Specify a format with command operators, similar to text_format, but from a DataSource.
+    DataSource text_format_source = 5;
```
Comment (qiwzhang) [inline api/envoy/config/core/v3/substitution_format_string.proto:69]: It seems that this should not be the place to add DataSource
1) why it only apply to text_format, why not json_format
2) data_source has `inline` which is same as `text_format` here.

Is it possible to specify in the file_access_log config to DataSource the whole SubstitutionFormatString?

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: This only applies to text_formart because json_format is, as I understand it, a convenience config for writing structured config inline to represent json text. For example:

```
json_format:
  x: 1
  y: 2
```

To supply json as text_format_source, one would write JSON text into the source and set content_type appropriately. Does this match your understanding? If we were to support a DataSource for json_format, what would be your preferred way of using it?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: That's a good question and I realize now that I forgot to call out the original motivation for this PR. I linked https://github.com/envoyproxy/envoy/pull/14221 as context.

The idea here is that we want a generic way to define HTTP response body rewrites. One interesting use case is to replace a non-empty HTTP body with an empty one, or one with a fixed string, like "denied".

One thing your comment exposes is that we're overloading the purpose of this object for generic rewrites and file_access_log formatting. There's probably an opportunity here to create some shared objects that don't cross concerns as much, making this less confusing.

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: Yeah as I said in #14221, I prefer just relax the text format validation. Will we see performance improvement of this formatter over the normal text_format?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: The question still exsit, what the rationale behind adding this as data_source, usually DataSource is used when the content is expected to be large, or it cannot fit in a protobuf string (i.e. binary). I don't see either of them apply?

**Review by lizan — CHANGES_REQUESTED** [review]
Please don't force push, it will screw up review history.

**Comment by esmet** [comment]
Sure, makes sense.

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: Great. I relax the format validation and remove fixed_text

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Agreed on the usage of DataSource. The rationale is that the text may be relatively large (e.g. a static html page) and may not be suitable for inline strings. The use case that comes to mind is a local reply or response map configuration that serves a variety of different html pages for different response codes.

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: I see that makes sense.

If that's the case, I think we want to deprecate `text_format` and use `text_format_source` as alternative, since DataSource has inline_string, then we should leave `fixed_text`.

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: > Is it possible to specify in the file_access_log config to DataSource the whole SubstitutionFormatString?

I don't see a reason it would be impossible. but for log format that's relatively small, there's no point to have it in a file.

Comment (qiwzhang) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: If we go with this way,  someone wants to read json_format from filter,  we would need to add json_format_source.  This will be bad.

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (qiwzhang) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: This is polluting SubstitutionFormatString class.  I see 1) `DataSource` of `text_format` and 2) `fixed_text` should not be part of this SubstitutionFormatString class.  They should be in the ResponseMapper message in https://github.com/envoyproxy/envoy/pull/14221

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Someone can still use `json_format` if they want to specify an inline representation of their JSON for filter config. If they want to specify a large JSON string from a DataSource, they'd be able to use `text_format_source` with the JSON in the file and then set `content_type` to `application/json`. I don't think `json_format_source` is necessary. What do you think?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: While I think it could be correct to specify a `DataSource` field in the `ResponseMapper` message in #14221, I also think it's reasonable and correct to allow for the `text_format` to come from a `DataSource` on `SubstitutionFormatString`, since a `DataSource` supports basic strings (via `inline_string`). The semantics of `SubstitutionFormatString` remain the same, we just have a behavioral extension that allows it to come from a source, which is nice. What do you think about the tradeoff of having the additional behavior here versus having users of this class define their own DataSource field to replace it?

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: I think there's also a reasonable argument to be made that the implementation of a `SubstitutionFormatString` using a `DataSouce` field was really straightforward. To move the source up a layer of abstraction, we'd need the option of creating a format string by telling it where it is sourced from. This conflicts with the existing fields in the format string that tell it where it is sourced from (i.e. text_format and json_format).

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: > I see that makes sense.
> 
> If that's the case, I think we want to deprecate `text_format` and use `text_format_source` as alternative, since DataSource has inline_string, then we should leave `fixed_text`.

I agree that if we go this route, then long term we do not need text_format since the DataSource variant supports inline strings via `inline_string`. If we did not relax the validation in `inline_string` to accept an empty string (it does not currently) then we'd need to indeed add `fixed_text` to support that case.

Comment (qiwzhang) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: I see.  I am fine with replacing `text_format` with `text_format_source`.
For `fixed_text`, it seems that its only use case is empty string since data_source.inline_string doesn't allow it.   Maybe change this field to `bool  empty`?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (qiwzhang) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: It is an interesting idea of replacing `json_format` with `text_format`.

**Review by qiwzhang — DISMISSED** [review]

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: I'm ok with a field for `empty`, though I think have a slight preference for `empty_format` which represents an empty format, in the same way that `text_format` represents the (non-empty) text format. @lizan originally commented on that here https://github.com/envoyproxy/envoy/pull/14221#discussion_r536029973 and suggested relaxing the constraints on an existing field, though I'm more of a fan of adding a new explicit field and preserving all existing behaviors. There could be some control plain implementations that depend on the existing field validation. I don't know how strong of an argument that is, though.

I'll vote for `bool empty_format`. @lizan / @qiwzhang final thoughts/votes?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 1 above)
Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: Now that I think about it, though, `json_format` strings are escaped for json while `text_format` strings are not. That's fairly important and prevents us from totally getting rid of the `json_format` field in favor of `text_format` until we have a way to guarantee that json-as-text can be escaped nicely. Maybe we could add/extend the command operator(s) to support escaping?

File: api/envoy/config/core/v3/substitution_format_string.proto (hunk 2 above)
Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Taking a step back, if you want empty, why would you specify a [`body_format`](https://github.com/envoyproxy/envoy/pull/14221/files#diff-8d3ae8065707ddf8386a198222c3872ef85eeaf282fabeacbe7a15b37da3d501R88), the field itself can be null?

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: I feel like a null SubstitutionFormatString could be interpreted as "no substitution will take place" as opposed to "substitute with nothing". What do you think?

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: > I feel like a null SubstitutionFormatString could be interpreted as "no substitution will take place" as opposed to "substitute with nothing". What do you think?

It depends on the context where this is used, but that seems natural.

So, after thinking a bit, my vote will be relaxing the constraint on what we have for `text_format` and/or `inline_string`/`inline_bytes`, because you can also have an empty file be specified as DataSource to make it empty, so that way it is more consistent.

Comment (qiwzhang) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: I sided with @lizan .  Adding extra fields may cause confusion, e.g. fixed text has to be specified in `text_format_source.inline_string`, but an empty one has to be specified in `empty_format`.

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Sounds good. I will relax the constraint on `text_format` and remove `fixed_text`.

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Let's also deprecate `text_format` in favor of `text_format_source`.

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Can do. For deprecations, is it appropriate to mark a field deprecated in `v3` and remove it in `v4alpha` or is there a different process?

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Yes, just mark `text_format` field deprecated and run the proto_format script, it will take care of the rest.

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Thanks. I should relax the field validation for `inline_string` on `DataSource` too, right?

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: That will have larger impact and I think we should do it in a separate PR, at least without it we can unblock #14221.

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:69]: Great. Thanks!

File: test/common/local_reply/local_reply_test.cc [hunk 3]
```diff
@@ -163,6 +163,24 @@ TEST_F(LocalReplyTest, TestMapperRewrite) {
...
+        inline_string: ""
+    - filter:
+        status_code_filter:
+          comparison:
+            op: EQ
+            value:
+              default_value: 404
+              runtime_key: key_b
+      body_format_override:
+        text_format: ""
```
Comment (qiwzhang) [inline test/common/local_reply/local_reply_test.cc:183]: Isn't this removed?

**Review by qiwzhang — CHANGES_REQUESTED** [review]
Need to fix tests.

File: test/common/local_reply/local_reply_test.cc (hunk 3 above)
Comment (esmet) [reply test/common/local_reply/local_reply_test.cc:183]: it's now deprecated, but not removed. since I removed the validation, I wanted to make sure there was a test case to cover the empty string case.

Comment (qiwzhang) [reply test/common/local_reply/local_reply_test.cc:183]: I see. Thanks

**Review by qiwzhang — DISMISSED** [review]

**Comment by lizan** [comment]
can you check CI?

**Comment by esmet** [comment]
> can you check CI?

Looks like tests are failing since we deprecated `text_format`.

How about we postpone the deprecation to another PR? This will allow us to move forward. If we do deprecate in this PR, then we'll need to update all of the tests to use inline_string, but then we won't be able to test the empty string case because we haven't related validation there (yet).

So, maybe we should deprecate both `text_source` and relax the `inline_string` validation in the next PR. What do you think?

**Comment by lizan** [comment]
> How about we postpone the deprecation to another PR? This will allow us to move forward. If we do deprecate in this PR, then we'll need to update all of the tests to use inline_string, but then we won't be able to test the empty string case because we haven't related validation there (yet).

You can still test it by wrapping test name with `DEPRECATED_FEATURE_TEST`, see how other test does. Deprecating doesn't mean removing tests.

**Comment by esmet** [comment]
Cool let me try that

**Review by lizan — COMMENTED** [review]
Thanks, this code LGTM, can you add release note? Need a line in deprecated section.

**Review by qiwzhang — APPROVED** [review]

**Review by lizan — COMMENTED** [review]
LGTM moduolo one nit. Thank you!

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -91,6 +93,7 @@ New Features
 
 Deprecated
 ----------
+* formatter: :ref:`text_format <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.text_format>` is now deprecated in favor of :ref:`text_format_source <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.text_format_source>`. To migrate existing text format strings, use the :ref:`inline_string <envoy_v3_api_field_config.core.v3.DataSource.inline_string>` field.
```
Comment (lizan) [inline docs/root/version_history/current.rst:96]: nit: alphabetical order

Comment (esmet) [reply docs/root/version_history/current.rst:96]: Done!

**Review by lizan — APPROVED** [review]
